File: funcscript/parser/syntax/FuncScriptParser.GetInfixFunctionCall.cs
using FuncScript.Block;
using FuncScript.Model;
using System.Collections.Generic;

namespace FuncScript.Core
{
    public partial class FuncScriptParser
    {
        static ExpressionBlockResult GetInfixFunctionCall(ParseContext context, int index)
        {
            int originalIndex = index;

            var leftResult = GetCallAndMemberAccess(context, index);
            if (leftResult.NextIndex == index)
            {
                return new ExpressionBlockResult(null, null, index);
            }

            var childNodes = new List<ParseNode>();
            ExpressionBlock expression = leftResult.Block;
            childNodes.Add(leftResult.ParseNode);
            
            index = SkipSpace(context, leftResult.NextIndex).NextIndex;

            bool foundAnyOperator = false;

            while (true)
            {
                var operatorResult = GetIdentifier(context, index, true);
                if (operatorResult.NextIndex == index)
                {
                    break;
                }

                var func = context.ReferenceProvider.Get(operatorResult.IdenLower);
                if (!(func is IFsFunction inf) || inf.CallType != CallType.Dual)
                {
                    break;
                }

                foundAnyOperator = true;

                childNodes.Add(operatorResult.ParseNode);

                index = SkipSpace(context, operatorResult.NextIndex).NextIndex;

                var rightResult = GetCallAndMemberAccess(context, index);
                if (rightResult.NextIndex == index)
                {
                    context.SyntaxErrors.Add(new SyntaxErrorData(
                        index,
                        0,
                        $"Right side operand expected for {operatorResult.Iden}"
                    ));
                    return new ExpressionBlockResult(null, null, originalIndex);
                }

                childNodes.Add(rightResult.ParseNode);
                index = SkipSpace(context, rightResult.NextIndex).NextIndex;

                var operands = new List<ExpressionBlock> { expression, rightResult.Block };

                while (true)
                {
                    var literalResult = GetLiteralMatch(context, index, "~");
                    if (literalResult.NextIndex == index)
                    {
                        break;
                    }

                    index = SkipSpace(context, literalResult.NextIndex).NextIndex;
                    var moreOperandResult = GetCallAndMemberAccess(context, index);
                    if (moreOperandResult.NextIndex == index)
                    {
                        break;
                    }

                    childNodes.Add(moreOperandResult.ParseNode);
                    index = SkipSpace(context, moreOperandResult.NextIndex).NextIndex;

                    operands.Add(moreOperandResult.Block);
                }

                var callExpr = new FunctionCallExpression
                {
                    Function = new LiteralBlock(func),
                    Parameters = operands.ToArray()
                };

                expression = callExpr;
            }

            if (!foundAnyOperator)
            {
                return leftResult;
            }

            int startPos = childNodes[0].Pos;
            var lastNode = childNodes[childNodes.Count - 1];
            int length = (lastNode.Pos + lastNode.Length) - startPos;
            
            var infixNode = new ParseNode(
                ParseNodeType.InfixExpression,
                startPos,
                length,
                childNodes
            );

            return new ExpressionBlockResult(expression, infixNode, index);
        }
    }
}

File: ts/core/src/parser/syntax/FuncScriptParser.GetInfixFunctionCall.ts
import { ParseContext, ParseNode, ParseNodeType, SyntaxErrorData } from "../FuncScriptParser.Main";
import { GetCallAndMemberAccess } from "./FuncScriptParser.GetCallAndMemberAccess";
import { GetIdentifier } from "./FuncScriptParser.GetIdentifier";
import { GetLiteralMatch } from "./FuncScriptParser.GetLiteralMatch";
import { SkipSpace } from "./FuncScriptParser.SkipSpace";

export function GetInfixFunctionCall(context: ParseContext, index: number): { ParseNode: ParseNode | null, NextIndex: number } {
    const originalIndex = index; 
    const childNodes: ParseNode[] = [];

    const leftResult = GetCallAndMemberAccess(context, index);
    if (leftResult.NextIndex === index) {
        return { ParseNode: null, NextIndex: index };
    }

    let parseNode = leftResult.ParseNode;
    childNodes.push(parseNode!);

    let i = SkipSpace(context, leftResult.NextIndex).NextIndex;
    let foundAnyOperator = false;

    while (true) {
        const operatorResult = GetIdentifier(context, i, true);
        if (operatorResult.NextIndex === i) {
            break; 
        }


        foundAnyOperator = true; 
        childNodes.push(operatorResult.ParseNode!);
        i = SkipSpace(context, operatorResult.NextIndex).NextIndex;

        const rightResult = GetCallAndMemberAccess(context, i);
        if (rightResult.NextIndex === i) {
            context.SyntaxErrors.push(new SyntaxErrorData(
                i, 
                0, 
                `Right side operand expected for ${operatorResult.Iden}`
            ));
            return { ParseNode: null, NextIndex: originalIndex };
        }

        childNodes.push(rightResult.ParseNode!);
        i = SkipSpace(context, rightResult.NextIndex).NextIndex;

        while (true) {
            const literalMatchResult = GetLiteralMatch(context, i, "~");
            if (literalMatchResult.NextIndex === i) break;

            i = SkipSpace(context, literalMatchResult.NextIndex).NextIndex;
            const moreOperandResult = GetCallAndMemberAccess(context, i);
            if (moreOperandResult.NextIndex === i) break;

            i = SkipSpace(context, moreOperandResult.NextIndex).NextIndex;
            childNodes.push(moreOperandResult.ParseNode!);
        }
    }

    if (!foundAnyOperator) {
        return leftResult; 
    }

    parseNode = new ParseNode(
        ParseNodeType.InfixExpression, 
        childNodes[0].Pos, 
        childNodes[childNodes.length - 1].Pos + childNodes[childNodes.length - 1].Length - childNodes[0].Pos, 
        childNodes
    );

    return { ParseNode: parseNode, NextIndex: i };
}